File: python/diagnyx/feedback.py
"""
User Feedback Collection Module

Provides methods for collecting end-user feedback on LLM responses.
Feedback is linked to traces for analysis and fine-tuning.

Usage:
    from diagnyx import Diagnyx

    diagnyx = Diagnyx(api_key="...")

    # Submit thumbs up feedback
    diagnyx.feedback.thumbs_up(trace_id="trace_123")

    # Submit rating feedback
    diagnyx.feedback.rating(trace_id="trace_123", value=4)

    # Submit text feedback with tags
    diagnyx.feedback.text(
        trace_id="trace_123",
        comment="Great response!",
        tags=["accurate", "helpful"]
    )

    # Submit correction
    diagnyx.feedback.correction(
        trace_id="trace_123",
        correction="The capital of France is Paris, not Lyon.",
    )
"""

from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional
import httpx


class FeedbackType(str, Enum):
    """Types of feedback that can be submitted."""

    THUMBS_UP = "thumbs_up"
    THUMBS_DOWN = "thumbs_down"
    RATING = "rating"
    TEXT = "text"
    CORRECTION = "correction"
    FLAG = "flag"


class FeedbackSentiment(str, Enum):
    """Sentiment classification of feedback."""

    POSITIVE = "positive"
    NEGATIVE = "negative"
    NEUTRAL = "neutral"


@dataclass
class Feedback:
    """Represents a feedback record."""

    id: str
    trace_id: str
    feedback_type: FeedbackType
    sentiment: FeedbackSentiment
    value: Optional[int] = None
    comment: Optional[str] = None
    correction: Optional[str] = None
    tags: List[str] = None
    metadata: Dict[str, Any] = None
    user_id: Optional[str] = None
    session_id: Optional[str] = None
    created_at: Optional[datetime] = None
    span_id: Optional[str] = None

    def __post_init__(self):
        if self.tags is None:
            self.tags = []
        if self.metadata is None:
            self.metadata = {}


@dataclass
class FeedbackSummary:
    """Summary analytics for feedback."""

    total_feedback: int
    positive_count: int
    negative_count: int
    neutral_count: int
    positive_rate: float
    average_rating: float
    feedback_by_type: Dict[str, int]
    feedback_by_tag: Dict[str, int]


class FeedbackClient:
    """
    Client for submitting and managing user feedback.

    Feedback is linked to traces and can be used for:
    - Monitoring user satisfaction
    - Identifying problematic responses
    - Collecting data for fine-tuning
    - Quality assurance
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.diagnyx.ai",
        organization_id: Optional[str] = None,
        project_id: Optional[str] = None,
        timeout: float = 30.0,
    ):
        """
        Initialize the feedback client.

        Args:
            api_key: Diagnyx API key
            base_url: API base URL
            organization_id: Organization ID
            project_id: Project ID
            timeout: Request timeout in seconds
        """
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.organization_id = organization_id
        self.project_id = project_id
        self.timeout = timeout
        self._client = httpx.Client(
            base_url=self.base_url,
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
            },
            timeout=timeout,
        )

    def thumbs_up(
        self,
        trace_id: str,
        *,
        span_id: Optional[str] = None,
        comment: Optional[str] = None,
        tags: Optional[List[str]] = None,
        user_id: Optional[str] = None,
        session_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Feedback:
        """
        Submit positive thumbs up feedback.

        Args:
            trace_id: The trace ID to attach feedback to
            span_id: Optional span ID for specific span feedback
            comment: Optional text comment
            tags: Optional tags for categorization
            user_id: Optional anonymized user identifier
            session_id: Optional session identifier
            metadata: Optional additional metadata

        Returns:
            The created Feedback object
        """
        return self._submit(
            trace_id=trace_id,
            feedback_type=FeedbackType.THUMBS_UP,
            span_id=span_id,
            comment=comment,
            tags=tags,
            user_id=user_id,
            session_id=session_id,
            metadata=metadata,
        )

    def thumbs_down(
        self,
        trace_id: str,
        *,
        span_id: Optional[str] = None,
        comment: Optional[str] = None,
        tags: Optional[List[str]] = None,
        user_id: Optional[str] = None,
        session_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Feedback:
        """
        Submit negative thumbs down feedback.

        Args:
            trace_id: The trace ID to attach feedback to
            span_id: Optional span ID for specific span feedback
            comment: Optional text comment explaining the issue
            tags: Optional tags for categorization (e.g., ["inaccurate", "off-topic"])
            user_id: Optional anonymized user identifier
            session_id: Optional session identifier
            metadata: Optional additional metadata

        Returns:
            The created Feedback object
        """
        return self._submit(
            trace_id=trace_id,
            feedback_type=FeedbackType.THUMBS_DOWN,
            span_id=span_id,
            comment=comment,
            tags=tags,
            user_id=user_id,
            session_id=session_id,
            metadata=metadata,
        )

    def rating(
        self,
        trace_id: str,
        value: int,
        *,
        span_id: Optional[str] = None,
        comment: Optional[str] = None,
        tags: Optional[List[str]] = None,
        user_id: Optional[str] = None,
        session_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Feedback:
        """
        Submit a numeric rating (1-5 stars).

        Args:
            trace_id: The trace ID to attach feedback to
            value: Rating value (1-5)
            span_id: Optional span ID for specific span feedback
            comment: Optional text comment
            tags: Optional tags for categorization
            user_id: Optional anonymized user identifier
            session_id: Optional session identifier
            metadata: Optional additional metadata

        Returns:
            The created Feedback object

        Raises:
            ValueError: If value is not between 1 and 5
        """
        if not 1 <= value <= 5:
            raise ValueError("Rating value must be between 1 and 5")

        return self._submit(
            trace_id=trace_id,
            feedback_type=FeedbackType.RATING,
            value=value,
            span_id=span_id,
            comment=comment,
            tags=tags,
            user_id=user_id,
            session_id=session_id,
            metadata=metadata,
        )

    def text(
        self,
        trace_id: str,
        comment: str,
        *,
        span_id: Optional[str] = None,
        tags: Optional[List[str]] = None,
        user_id: Optional[str] = None,
        session_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Feedback:
        """
        Submit text feedback/comment.

        Args:
            trace_id: The trace ID to attach feedback to
            comment: The text feedback
            span_id: Optional span ID for specific span feedback
            tags: Optional tags for categorization
            user_id: Optional anonymized user identifier
            session_id: Optional session identifier
            metadata: Optional additional metadata

        Returns:
            The created Feedback object
        """
        return self._submit(
            trace_id=trace_id,
            feedback_type=FeedbackType.TEXT,
            comment=comment,
            span_id=span_id,
            tags=tags,
            user_id=user_id,
            session_id=session_id,
            metadata=metadata,
        )

    def correction(
        self,
        trace_id: str,
        correction: str,
        *,
        span_id: Optional[str] = None,
        comment: Optional[str] = None,
        tags: Optional[List[str]] = None,
        user_id: Optional[str] = None,
        session_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Feedback:
        """
        Submit a correction for the response.

        Corrections are used for fine-tuning - they provide the "correct"
        response that should have been generated.

        Args:
            trace_id: The trace ID to attach feedback to
            correction: The corrected response text
            span_id: Optional span ID for specific span feedback
            comment: Optional comment explaining the correction
            tags: Optional tags for categorization
            user_id: Optional anonymized user identifier
            session_id: Optional session identifier
            metadata: Optional additional metadata

        Returns:
            The created Feedback object
        """
        return self._submit(
            trace_id=trace_id,
            feedback_type=FeedbackType.CORRECTION,
            correction=correction,
            span_id=span_id,
            comment=comment,
            tags=tags,
            user_id=user_id,
            session_id=session_id,
            metadata=metadata,
        )

    def flag(
        self,
        trace_id: str,
        *,
        reason: Optional[str] = None,
        span_id: Optional[str] = None,
        tags: Optional[List[str]] = None,
        user_id: Optional[str] = None,
        session_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Feedback:
        """
        Flag a response for review.

        Use this for responses that need human review (e.g., potentially
        harmful content, policy violations, etc.)

        Args:
            trace_id: The trace ID to flag
            reason: Optional reason for flagging
            span_id: Optional span ID for specific span
            tags: Optional tags (e.g., ["harmful", "policy-violation"])
            user_id: Optional anonymized user identifier
            session_id: Optional session identifier
            metadata: Optional additional metadata

        Returns:
            The created Feedback object
        """
        return self._submit(
            trace_id=trace_id,
            feedback_type=FeedbackType.FLAG,
            span_id=span_id,
            comment=reason,
            tags=tags,
            user_id=user_id,
            session_id=session_id,
            metadata=metadata,
        )

    def _submit(
        self,
        trace_id: str,
        feedback_type: FeedbackType,
        *,
        value: Optional[int] = None,
        span_id: Optional[str] = None,
        comment: Optional[str] = None,
        correction: Optional[str] = None,
        tags: Optional[List[str]] = None,
        user_id: Optional[str] = None,
        session_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Feedback:
        """
        Internal method to submit feedback.

        Args:
            trace_id: The trace ID
            feedback_type: Type of feedback
            value: Optional numeric value
            span_id: Optional span ID
            comment: Optional comment
            correction: Optional correction text
            tags: Optional tags
            user_id: Optional user ID
            session_id: Optional session ID
            metadata: Optional metadata

        Returns:
            The created Feedback object
        """
        payload = {
            "traceId": trace_id,
            "feedbackType": feedback_type.value,
        }

        if span_id:
            payload["spanId"] = span_id
        if value is not None:
            payload["value"] = value
        if comment:
            payload["comment"] = comment
        if correction:
            payload["correction"] = correction
        if tags:
            payload["tags"] = tags
        if user_id:
            payload["userId"] = user_id
        if session_id:
            payload["sessionId"] = session_id
        if metadata:
            payload["metadata"] = metadata

        response = self._client.post("/api/v1/feedback", json=payload)
        response.raise_for_status()

        data = response.json()

        return Feedback(
            id=data.get("id"),
            trace_id=trace_id,
            feedback_type=feedback_type,
            sentiment=FeedbackSentiment(data.get("sentiment", "neutral")),
            value=value,
            comment=comment,
            correction=correction,
            tags=tags or [],
            metadata=metadata or {},
            user_id=user_id,
            session_id=session_id,
            span_id=span_id,
            created_at=datetime.fromisoformat(data["createdAt"].replace("Z", "+00:00"))
            if "createdAt" in data
            else None,
        )

    def get_summary(
        self,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
    ) -> FeedbackSummary:
        """
        Get feedback analytics summary.

        Args:
            start_date: Optional start date filter
            end_date: Optional end date filter

        Returns:
            FeedbackSummary with analytics data
        """
        params = {}
        if start_date:
            params["startDate"] = start_date.isoformat()
        if end_date:
            params["endDate"] = end_date.isoformat()

        response = self._client.get("/api/v1/feedback/summary", params=params)
        response.raise_for_status()

        data = response.json()

        return FeedbackSummary(
            total_feedback=data["totalFeedback"],
            positive_count=data["positiveCount"],
            negative_count=data["negativeCount"],
            neutral_count=data["neutralCount"],
            positive_rate=data["positiveRate"],
            average_rating=data["averageRating"],
            feedback_by_type=data["feedbackByType"],
            feedback_by_tag=data["feedbackByTag"],
        )

    def list(
        self,
        limit: int = 50,
        offset: int = 0,
        feedback_type: Optional[FeedbackType] = None,
        sentiment: Optional[FeedbackSentiment] = None,
        tag: Optional[str] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
    ) -> List[Feedback]:
        """
        List feedback with filters.

        Args:
            limit: Maximum number of results
            offset: Offset for pagination
            feedback_type: Filter by feedback type
            sentiment: Filter by sentiment
            tag: Filter by tag
            start_date: Filter by start date
            end_date: Filter by end date

        Returns:
            List of Feedback objects
        """
        params = {"limit": limit, "offset": offset}
        if feedback_type:
            params["feedbackType"] = feedback_type.value
        if sentiment:
            params["sentiment"] = sentiment.value
        if tag:
            params["tag"] = tag
        if start_date:
            params["startDate"] = start_date.isoformat()
        if end_date:
            params["endDate"] = end_date.isoformat()

        response = self._client.get("/api/v1/feedback", params=params)
        response.raise_for_status()

        data = response.json()

        return [
            Feedback(
                id=item["id"],
                trace_id=item["traceId"],
                feedback_type=FeedbackType(item["feedbackType"]),
                sentiment=FeedbackSentiment(item["sentiment"]),
                value=item.get("value"),
                comment=item.get("comment"),
                correction=item.get("correction"),
                tags=item.get("tags", []),
                metadata=item.get("metadata", {}),
                user_id=item.get("userId"),
                session_id=item.get("sessionId"),
                span_id=item.get("spanId"),
                created_at=datetime.fromisoformat(
                    item["createdAt"].replace("Z", "+00:00")
                )
                if "createdAt" in item
                else None,
            )
            for item in data.get("data", [])
        ]

    def close(self):
        """Close the HTTP client."""
        self._client.close()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

File: go/feedback.go
package diagnyx

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

// FeedbackType represents the type of feedback
type FeedbackType string

const (
	FeedbackTypeThumbsUp   FeedbackType = "thumbs_up"
	FeedbackTypeThumbsDown FeedbackType = "thumbs_down"
	FeedbackTypeRating     FeedbackType = "rating"
	FeedbackTypeText       FeedbackType = "text"
	FeedbackTypeCorrection FeedbackType = "correction"
	FeedbackTypeFlag       FeedbackType = "flag"
)

// FeedbackSentiment represents the sentiment classification
type FeedbackSentiment string

const (
	FeedbackSentimentPositive FeedbackSentiment = "positive"
	FeedbackSentimentNegative FeedbackSentiment = "negative"
	FeedbackSentimentNeutral  FeedbackSentiment = "neutral"
)

// FeedbackOptions contains optional parameters for feedback submission
type FeedbackOptions struct {
	// Optional span ID for specific span feedback
	SpanID string
	// Optional text comment
	Comment string
	// Tags for categorization
	Tags []string
	// Additional metadata
	Metadata map[string]interface{}
	// Anonymized user identifier
	UserID string
	// Session identifier
	SessionID string
}

// Feedback represents a feedback record
type Feedback struct {
	ID           string                 `json:"id"`
	TraceID      string                 `json:"traceId"`
	FeedbackType FeedbackType           `json:"feedbackType"`
	Sentiment    FeedbackSentiment      `json:"sentiment"`
	Rating       *int                   `json:"rating,omitempty"`
	Comment      string                 `json:"comment,omitempty"`
	Correction   string                 `json:"correction,omitempty"`
	Tags         []string               `json:"tags"`
	Metadata     map[string]interface{} `json:"metadata"`
	UserID       string                 `json:"userId,omitempty"`
	SessionID    string                 `json:"sessionId,omitempty"`
	SpanID       string                 `json:"spanId,omitempty"`
	CreatedAt    time.Time              `json:"createdAt"`
}

// FeedbackSummary contains feedback analytics
type FeedbackSummary struct {
	TotalFeedback  int            `json:"totalFeedback"`
	PositiveCount  int            `json:"positiveCount"`
	NegativeCount  int            `json:"negativeCount"`
	NeutralCount   int            `json:"neutralCount"`
	PositiveRate   float64        `json:"positiveRate"`
	AverageRating  float64        `json:"averageRating"`
	FeedbackByType map[string]int `json:"feedbackByType"`
	FeedbackByTag  map[string]int `json:"feedbackByTag"`
}

// ListFeedbackOptions contains parameters for listing feedback
type ListFeedbackOptions struct {
	Limit        int
	Offset       int
	FeedbackType FeedbackType
	Sentiment    FeedbackSentiment
	Tag          string
	StartDate    *time.Time
	EndDate      *time.Time
}

// ListFeedbackResult contains the paginated feedback list
type ListFeedbackResult struct {
	Data   []Feedback `json:"data"`
	Total  int        `json:"total"`
	Limit  int        `json:"limit"`
	Offset int        `json:"offset"`
}

// FeedbackClient handles feedback operations
type FeedbackClient struct {
	apiKey         string
	baseURL        string
	organizationID string
	maxRetries     int
	debug          bool
	httpClient     *http.Client
}

// NewFeedbackClient creates a new feedback client
func NewFeedbackClient(apiKey, organizationID string, opts ...FeedbackClientOption) *FeedbackClient {
	c := &FeedbackClient{
		apiKey:         apiKey,
		baseURL:        "https://api.diagnyx.io",
		organizationID: organizationID,
		maxRetries:     3,
		debug:          false,
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}

	for _, opt := range opts {
		opt(c)
	}

	return c
}

// FeedbackClientOption configures a FeedbackClient
type FeedbackClientOption func(*FeedbackClient)

// WithFeedbackBaseURL sets the base URL
func WithFeedbackBaseURL(url string) FeedbackClientOption {
	return func(c *FeedbackClient) {
		c.baseURL = url
	}
}

// WithFeedbackMaxRetries sets the max retries
func WithFeedbackMaxRetries(retries int) FeedbackClientOption {
	return func(c *FeedbackClient) {
		c.maxRetries = retries
	}
}

// WithFeedbackDebug enables debug mode
func WithFeedbackDebug(debug bool) FeedbackClientOption {
	return func(c *FeedbackClient) {
		c.debug = debug
	}
}

// ThumbsUp submits positive feedback
func (c *FeedbackClient) ThumbsUp(traceID string, opts *FeedbackOptions) (*Feedback, error) {
	return c.submit(traceID, FeedbackTypeThumbsUp, nil, "", "", opts)
}

// ThumbsDown submits negative feedback
func (c *FeedbackClient) ThumbsDown(traceID string, opts *FeedbackOptions) (*Feedback, error) {
	return c.submit(traceID, FeedbackTypeThumbsDown, nil, "", "", opts)
}

// Rating submits a numeric rating (1-5)
func (c *FeedbackClient) Rating(traceID string, value int, opts *FeedbackOptions) (*Feedback, error) {
	if value < 1 || value > 5 {
		return nil, fmt.Errorf("rating value must be between 1 and 5")
	}
	return c.submit(traceID, FeedbackTypeRating, &value, "", "", opts)
}

// Text submits text feedback
func (c *FeedbackClient) Text(traceID, comment string, opts *FeedbackOptions) (*Feedback, error) {
	return c.submit(traceID, FeedbackTypeText, nil, comment, "", opts)
}

// Correction submits a correction for fine-tuning
func (c *FeedbackClient) Correction(traceID, correction string, opts *FeedbackOptions) (*Feedback, error) {
	return c.submit(traceID, FeedbackTypeCorrection, nil, "", correction, opts)
}

// Flag flags a response for review
func (c *FeedbackClient) Flag(traceID string, reason string, opts *FeedbackOptions) (*Feedback, error) {
	return c.submit(traceID, FeedbackTypeFlag, nil, reason, "", opts)
}

func (c *FeedbackClient) submit(traceID string, feedbackType FeedbackType, rating *int, comment, correction string, opts *FeedbackOptions) (*Feedback, error) {
	if opts == nil {
		opts = &FeedbackOptions{}
	}

	payload := map[string]interface{}{
		"traceId":      traceID,
		"feedbackType": feedbackType,
	}

	if opts.SpanID != "" {
		payload["spanId"] = opts.SpanID
	}
	if rating != nil {
		payload["rating"] = *rating
	}
	if comment != "" {
		payload["comment"] = comment
	} else if opts.Comment != "" {
		payload["comment"] = opts.Comment
	}
	if correction != "" {
		payload["correction"] = correction
	}
	if len(opts.Tags) > 0 {
		payload["tags"] = opts.Tags
	}
	if opts.Metadata != nil {
		payload["metadata"] = opts.Metadata
	}
	if opts.UserID != "" {
		payload["userId"] = opts.UserID
	}
	if opts.SessionID != "" {
		payload["sessionId"] = opts.SessionID
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	var result Feedback
	err = c.request("POST", "/api/v1/feedback", body, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// List retrieves feedback with filters
func (c *FeedbackClient) List(opts *ListFeedbackOptions) (*ListFeedbackResult, error) {
	if opts == nil {
		opts = &ListFeedbackOptions{}
	}

	params := url.Values{}
	if opts.Limit > 0 {
		params.Set("limit", fmt.Sprintf("%d", opts.Limit))
	}
	if opts.Offset > 0 {
		params.Set("offset", fmt.Sprintf("%d", opts.Offset))
	}
	if opts.FeedbackType != "" {
		params.Set("feedbackType", string(opts.FeedbackType))
	}
	if opts.Sentiment != "" {
		params.Set("sentiment", string(opts.Sentiment))
	}
	if opts.Tag != "" {
		params.Set("tag", opts.Tag)
	}
	if opts.StartDate != nil {
		params.Set("startDate", opts.StartDate.Format(time.RFC3339))
	}
	if opts.EndDate != nil {
		params.Set("endDate", opts.EndDate.Format(time.RFC3339))
	}

	path := fmt.Sprintf("/api/v1/organizations/%s/feedback", c.organizationID)
	if len(params) > 0 {
		path += "?" + params.Encode()
	}

	var result ListFeedbackResult
	err := c.request("GET", path, nil, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// GetSummary retrieves feedback analytics
func (c *FeedbackClient) GetSummary(startDate, endDate *time.Time) (*FeedbackSummary, error) {
	params := url.Values{}
	if startDate != nil {
		params.Set("startDate", startDate.Format(time.RFC3339))
	}
	if endDate != nil {
		params.Set("endDate", endDate.Format(time.RFC3339))
	}

	path := fmt.Sprintf("/api/v1/organizations/%s/feedback/analytics", c.organizationID)
	if len(params) > 0 {
		path += "?" + params.Encode()
	}

	var result FeedbackSummary
	err := c.request("GET", path, nil, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// GetForTrace retrieves feedback for a specific trace
func (c *FeedbackClient) GetForTrace(traceID string) ([]Feedback, error) {
	path := fmt.Sprintf("/api/v1/organizations/%s/feedback/trace/%s", c.organizationID, traceID)

	var result []Feedback
	err := c.request("GET", path, nil, &result)
	if err != nil {
		return nil, err
	}

	return result, nil
}

func (c *FeedbackClient) request(method, path string, body []byte, result interface{}) error {
	var lastErr error

	for attempt := 0; attempt < c.maxRetries; attempt++ {
		req, err := http.NewRequest(method, c.baseURL+path, bytes.NewReader(body))
		if err != nil {
			return fmt.Errorf("failed to create request: %w", err)
		}

		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Authorization", "Bearer "+c.apiKey)

		resp, err := c.httpClient.Do(req)
		if err != nil {
			lastErr = err
			c.log("Attempt %d failed: %v", attempt+1, err)
			time.Sleep(time.Duration(1<<attempt) * time.Second)
			continue
		}
		defer resp.Body.Close()

		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			if result != nil {
				if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
					return fmt.Errorf("failed to decode response: %w", err)
				}
			}
			return nil
		}

		lastErr = fmt.Errorf("HTTP %d", resp.StatusCode)
		c.log("Attempt %d failed: %v", attempt+1, lastErr)

		if resp.StatusCode >= 400 && resp.StatusCode < 500 {
			// Don't retry client errors
			return lastErr
		}

		time.Sleep(time.Duration(1<<attempt) * time.Second)
	}

	return lastErr
}

func (c *FeedbackClient) log(format string, args ...interface{}) {
	if c.debug {
		fmt.Printf("[Diagnyx Feedback] "+format+"\n", args...)
	}
}
